File: src/main/java/xyz/d1snin/corby/event/reactions/ReactionEvent.java
/*
 * BSD 3-Clause License
 *
 * Copyright (c) 2021, Corby
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package xyz.d1snin.corby.event.reactions;

import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.events.GenericEvent;
import net.dv8tion.jda.api.events.message.react.MessageReactionAddEvent;
import xyz.d1snin.corby.Corby;
import xyz.d1snin.corby.event.Listener;

public abstract class ReactionEvent extends Listener {

  protected String emoji;

  public ReactionEvent() {
    this.event = MessageReactionAddEvent.class;
  }

  protected abstract void performReaction(MessageReactionAddEvent event, Message msg);

  @Override
  public void perform(GenericEvent event) {
    MessageReactionAddEvent thisEvent = (MessageReactionAddEvent) event;

    if (emoji.equals(thisEvent.getReaction().getReactionEmote().getName())) {
      Corby.getService()
          .execute(() -> performReaction(thisEvent, thisEvent.retrieveMessage().complete()));
    }
  }
}

File: src/main/kotlin/xyz/d1snin/corby/event/reactions/ReactionEvent.kt
/*
 * BSD 3-Clause License, Copyright (c) 2021, Corby and contributors, All rights reserved.
 */

package xyz.d1snin.corby.event.reactions

import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.events.message.guild.react.GuildMessageReactionAddEvent
import xyz.d1snin.corby.event.EventListener

abstract class ReactionEvent(protected val emoji: String) : EventListener<GuildMessageReactionAddEvent>() {
    protected lateinit var msg: Message
    private lateinit var action: GuildMessageReactionAddEvent.() -> Unit

    protected fun perform(block: GuildMessageReactionAddEvent.() -> Unit) {
        action = block
    }

    init {
        execute {
            if (emoji == reaction.reactionEmote.name) {
                msg = retrieveMessage().complete()
                action(this)
            }
        }
    }
}